testng-core/src/main/java/org/testng/internal/ConfigurationGroupMethods.java

```diff
@@ -117,6 +130,7 @@ public List<ITestNGMethod> getAfterGroupMethodsForGroup(String group) {
   public void removeBeforeGroups(String[] groups) {
     for (String group : groups) {
       m_beforeGroupsMethods.remove(group);
+      beforeGroupsThatHaveAlreadyRun.get(group).countDown();
```

Comment:
If there are more than 1 `beforeGroups()` that would need to be run in parallel, wouldn't this `countDown()` cause them to run one after the other because the `for..each` is running in sequential order. Should we use `Arrays.stream().parallel()` instead so that we don't introduce this sequential behaviour ?

Reply:
countDown call does not block thread here, but unblocks other waiting threads. 
And, imo, using parallel stream can cause other bottleneck (due to commonPool usage)

Reply:
Ah! yeah. _facepalm moment_
My bad. That makes sense. The `await()` is what blocks and not the `countDown()`.